.github/workflows/ci.yml

```diff
@@ -0,0 +1,186 @@
+name: CI
+
+on:
+  push:
+  pull_request:
+    branches:
+      - main
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+env:
+  GO_VERSION: 1.18
+  CONSUL_VERSION: 1.12.1
+  TERRAFORM_VERSION: 1.2.2
+  GOTESTSUM_VERSION: 1.8.1
+
+jobs:
+  checks:
+    name: Checks
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Set up Go
+        uses: actions/setup-go@v3
+        with:
+          go-version: ${{ env.GO_VERSION }}
+
+      - name: Run Go checks
+        run: |
+          make go-fmt-check
+
+      - name: Run Terraform checks
+        run: |
+          make terraform-fmt-check
+
+      - name: Compile weekly tests
+        run: |
+          make compile-weekly-tests
+
+      - name: Verify go.mod and go.sum are correct
+        run: |
+          go mod tidy
+          git diff --quiet && exit 0
+          echo "please run 'go mod tidy' to ensure go.mod and go.sum are up to date"
+          exit 1
+
+  build:
+    name: Build
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Set up Go
+        uses: actions/setup-go@v3
+        with:
+          go-version: ${{ env.GO_VERSION }}
+
+      - name: Build
+        run: |
+          go build -ldflags "-w -s"
+
+  unit-and-integration:
+    name: Unit and Integration Tests
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Set up Go
+        uses: actions/setup-go@v3
+        with:
+          go-version: ${{ env.GO_VERSION }}
+
+      - name: Install Consul
```

Comment:
Is there some way to abstract these installs to a step or bash script, since we're using it in multiple places?

Reply:
Hi @hashi-derek, I switched to re-using actions that already exist and extracted out the other ones.

Reply:
@jwair That looks really nice